Homework2/Task1/Classes/GenericDiagonalMatrix.cs

```diff
@@ -0,0 +1,73 @@
+﻿namespace Task1.Classes
+{
+ class GenericDiagonalMatrix<T>
+ {
+  private T[] MatrixArray { get; set; }
+
+  public T this[int i, int j]
+  {
+   get
+   {
+    if (i < 0 || j < 0 ||
+        i >= Size || j >= Size)
+    {
+     throw new IndexOutOfRangeException();
+    }
+    if (i != j)
+    {
+     return default(T);
+    }
+    return MatrixArray[i];
+   }
+
+   set
+   {
+    if (i != j)
+    {
+     throw new Exception();
+    }
+    if (!EqualityComparer<T>.Default.Equals(value, MatrixArray[i]) &&
+        !EqualityComparer<T>.Default.Equals(value, default(T)))
+    {
+     OnElementChange?.Invoke(this, MatrixArray[i], value, i);
```

Comment:
Event should be triggered when element has been changed already, so swap the line with the code below